Migration guide from track1 to track2 (#22879)
Azure/azure-sdk-for-net: sdk/resourcemanager/Azure.ResourceManager/docs/MigrationGuide.md

# Migrating from old to preview management SDK.

There are several differences between the old sdk and this preview. Here's an example of how to create a Virtual Machine with both SDKs:

## Create a Virtual Machine example

### Import the namespaces
#### Old (Microsoft.Azure.Management._)
```C#
using Microsoft.Azure.Management.Compute;
using Microsoft.Azure.Management.Compute.Models;
using Microsoft.Azure.Management.Network;
using Microsoft.Azure.Management.Network.Models;
using Microsoft.Azure.Management.ResourceManager;
using Microsoft.Rest;
using System;
using System.Threading.Tasks;
```
#### New (Azure.Azure.ResourceManager._ Preview)
```C#
using Azure.Identity;

> using Azure.ResourceManager.Core;

Reviewer: Isn't the root namespace Azure.ResourceManager now?

Author: You're right, thanks for pointing that out

Author: Actually, I think it's `Azure.ResourceManager.Resources`

var networkClient = new NetworkManagementClient(credentials);
var managedServiceIdentityClient = new ManagedServiceIdentityClient(credentials);
```
#### New
```C#
var armClient = new ArmClient(new DefaultAzureCredential());
```
As you can see, authentication is now handled by Azure.Identity, and now just a single client is needed, from which you can get the `DefaultSubscription` and start managing your resources. 

### Create a Resource Group
#### Old
```C#
ServiceClientCredentials credentials = getMyCredentials(); 
var resourcesClient = new ResourceManagementClient(credentials);

string rgName = "QuickStartRG";
string location = "WestUS2";
resourcesClient.ResourceGroups.CreateOrUpdate(
    rgName,
    new ResourceGroup
    {
        Location = location,
        Tags = new Dictionary<string, string>() { { rgName, DateTime.UtcNow.ToString("u") } }
    });
```
#### New
```C#
var armClient = new ArmClient(new DefaultAzureCredential());
Subscription subscription = armClient.DefaultSubscription;
ResourceGroupContainer rgContainer = subscription.GetResourceGroups();

Location location = Location.WestUS2;
string rgName = "QuickStartRG";

> ResourceGroup resourceGroup = await rgContainer.Construct(location).CreateOrUpdateAsync(rgName)

Reviewer 2: Construct is removed for preview call Create directly and pass in a new ResourceGroupData(...)

```

> In the new libraries, the Resource Group is created trought a container that came from the subscription. Also, public locations are provided trought the `Location` object, but it can be specified as a string as well. Additionally, there's no need to create a `ResourceGroup` object to send parameters when creating a resource group, since tags can be added directly from the resource group variable and the location is specified in the `Construct()` method.

Reviewer: ```suggestion
In the new libraries, the Resource Group is created through a container that came from the subscription. Also, public locations are provided through the `Location` object, but can be specified using a string as well. Additionally, there's no need to create a `ResourceGroup` object to send parameters when creating a resource group, since tags can be added directly from the resource group variable and the location is specified in the `Construct()` method.
```

Reviewer: Actually, we probably want to say a couple of things here a bit differently.

First, the main difference is that the legacy client api represents all operations as flat, while the preview client API represents the hierarchy of resources, so that, for example, you can use a 'subscriptionContainer' to manage the resources in a particular subscription, you can use a resourceGroupContainer to manage the resources in a particular resource group, etc.

Second, it isn't quite true that you don't have to create a resource group object - rather, we provide some 'Construct' method helpers in the resource group container to help create common configurations of common resources, like resource group.  These make it easier to discover the correct way to create properly configured resources.

Reviewer: Third, we have provided some common classes to represent commonly-used constructs, like Location, and allow you to use them directly throughout the APIs, again, making it easier to discover how to properly configure resources.

### Create an Availability Set
#### Old
```C#
var inputAvailabilitySet = new AvailabilitySet
{
    Location = location,
    Tags = new Dictionary<string, string>()
    {
        { "RG", "rg" },
        { "testTag", "1" }
    },
    PlatformFaultDomainCount = 1,
    PlatformUpdateDomainCount = 1,
    Sku = new CM.Sku
    {
        Name = AvailabilitySetSkuTypes.Aligned 
    }
};
string aSetName = "quickstartvm_aSet";

var asCreateOrUpdateResponse = computeClient.AvailabilitySets.CreateOrUpdate(
    rgName,
    aSetName,
    inputAvailabilitySet
);

> string aSetID = $"/subscriptions/{computeClient.SubscriptionId}/resourceGroups/{rgName}/providers/Microsoft.Compute/availabilitySets/{aSetName}";

Reviewer: You should probably show accessing the resource id for the availability set in the preview example below:
``` c#
string asetId = aset.Id;
```

```
#### New
```C#
string vmName = "quickstartvm";

> var aset = await resourceGroup.GetAvailabilitySets().Construct("Aligned").CreateOrUpdateAsync(vmName + "_aSet");

Reviewer 2: same comment on construct, we probably need to change this in the [samples ](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/resourcemanager/Azure.ResourceManager/samples/Sample3_CreatingAVirtualNetwork.md)too.

### Create a Virtual Network and Subnet
#### Old
```C#
string vnetName = vmName + "_vnet";
string subnetName = "mySubnet";

var vnet = new VirtualNetwork()
{
    Location = location,
    AddressSpace = new AddressSpace()
    {
        AddressPrefixes = new List<string>()
                {
                    "10.0.0.0/16",
                }
    },
    DhcpOptions = !addDnsServer ? null : new DhcpOptions()
    {
        DnsServers = new List<string>()
                {
                    "10.1.1.1",
                    "10.1.2.4"
                }
    },
    Subnets = new List<Subnet>()
            {
                new Subnet()
                {
                    Name = subnetName,
                    AddressPrefix = "10.0.0.0/24",
                    PrivateEndpointNetworkPolicies = disablePEPolicies ? "Disabled" : null
                }
            }
};
var putVnetResponse = networkClient.VirtualNetworks.CreateOrUpdate(rgName, vnetName, vnet);
var subnetResponse = networkClient.Subnets.Get(rgName, vnetName, subnetName);
```
#### New
```C#
string vnetName = vmName + "_vnet";
string subnetName = "mySubnet";

> var vnet = await resourceGroup.GetVirtualNetworks().Construct("10.0.0.0/16").CreateOrUpdateAsync(vnetName);

Reviewer: I don't think you are setting dns servers in this example.

    },
    StorageProfile = new StorageProfile
    {
        ImageReference = imageRef,
        OsDisk = new OSDisk
        {
            Caching = CachingTypes.None,
            WriteAcceleratorEnabled = writeAcceleratorEnabled,
            CreateOption = DiskCreateOptionTypes.FromImage,
            Name = "test",
            Vhd =  null,
            ManagedDisk = new ManagedDiskParameters
            {
                StorageAccountType = osDiskStorageAccountType,
                DiskEncryptionSet = null,
            }
        },
        DataDisks = null,
    },
    OsProfile = new OSProfile
    {
        AdminUsername = "Foo12",
        AdminPassword = PLACEHOLDER,
        ComputerName = ComputerName
    }
};

inputVM.AvailabilitySet = new Microsoft.Azure.Management.Compute.Models.SubResource() { Id = aSetID };

CM.NetworkProfile vmNetworkProfile = new CM.NetworkProfile();
vmNetworkProfile.NetworkInterfaces = new List<NetworkInterfaceReference>
{
    new NetworkInterfaceReference
    {
        Id = nicResponse.Id
    }
};
inputVM.NetworkProfile = vmNetworkProfile;

VirtualMachine vm = VMcomputeClient.VirtualMachines.CreateOrUpdate(rgName, inputVM.Name, inputVM);
```

> #### New

Reviewer 3: this doc is great! besides guides on specific resource operation, how about add general operations migration guide, like paging, error handling, customized policy etc? similar like Java migration guide: https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/resourcemanager/docs/MIGRATION_GUIDE.md

    rgName,
    new ResourceGroup
    {
        Location = location,
        Tags = new Dictionary<string, string>() { { rgName, DateTime.UtcNow.ToString("u") } }
    });
```
#### New
```C#
Subscription subscription = armClient.DefaultSubscription;
ResourceGroupContainer rgContainer = subscription.GetResourceGroups();

Location location = Location.WestUS2;
string rgName = "QuickStartRG";

ResourceGroupData rgData = new ResourceGroupData(location);
ResourceGroup resourceGroup = await rgContainer.CreateOrUpdateAsync(rgName, rgData);
```
The main difference is that the previous libraries represent all operations as flat, while the new preview libraries respresents the hierarchy of resources. In that way, you can use a `subscriptionContainer` to manage the resources in a particular subscription. In this example, a `resourceGroupContainer` is used to manage the resources in a particular resource group. In the example above, a new resource group is created from a resourceGroupContainer. With that `ResourceGroup` you will be able to get the resource containers to manage all the resources that will be inside it, as it is shown in the next part of this guide.

> Another thing is that the new preview SDK provides some common classes to represent commonly-used constructs, like `Location`, and allow you to use them directly throughout the APIs, again, making it easier to discover how to properly configure resources.

Reviewer: ```suggestion
The new preview SDK also provides some common classes to represent commonly-used constructs, like `Location`, and allows you to use them directly throughout the APIs, making it easier to discover how to properly configure resources.
```

### Create an Availability Set
#### Old
```C#
AvailabilitySet inputAvailabilitySet = new AvailabilitySet
{
    Location = location,
    Tags = new Dictionary<string, string>()
    {
        { "RG", "rg" },
        { "testTag", "1" }
    },
    PlatformFaultDomainCount = 1,
    PlatformUpdateDomainCount = 1,
    Sku = new CM.Sku
    {
        Name = AvailabilitySetSkuTypes.Aligned 
    }
};
string aSetName = "quickstartvm_aSet";

AvailabilitySet asCreateOrUpdateResponse = computeClient.AvailabilitySets.CreateOrUpdate(rgName,aSetName,inputAvailabilitySet);
string aSetID = $"/subscriptions/{computeClient.SubscriptionId}/resourceGroups/{rgName}/providers/Microsoft.Compute/availabilitySets/{aSetName}";
```
#### New
```C#
string vmName = "quickstartvm";
AvailabilitySetData aSetData = new AvailabilitySetData(location);
AvailabilitySet aset = (await resourceGroup.GetAvailabilitySets().CreateOrUpdateAsync(vmName + "_aSet", aSetData)).Value;
string asetId = aset.Id;
```

> Parameters can be specified via the `AvailabilitySetData` object, in here, the basic default only requires the location. The availability set is created using  the `GetAvailabilitySets()` extension method instead of using another client. 

Reviewer: ```suggestion
Parameters can be specified via the `AvailabilitySetData` object, in here, the basic default only requires the location. The availability set is created using  the AvailabilitySetsContainer returned from the `GetAvailabilitySets()` extension method instead of using another client. 
```

VirtualNetwork vnet = new VirtualNetwork()
{
    Location = location,
    AddressSpace = new AddressSpace()
    {
        AddressPrefixes = new List<string>()
                {
                    "10.0.0.0/16",
                }
    },
    Subnets = new List<Subnet>()
            {
                new Subnet()
                {
                    Name = subnetName,
                    AddressPrefix = "10.0.0.0/24",
                }
            }
};
VirtualNetwork putVnetResponse = networkClient.VirtualNetworks.CreateOrUpdate(rgName, vnetName, vnet);
VirtualNetwork subnetResponse = networkClient.Subnets.Get(rgName, vnetName, subnetName);
```
#### New
```C#
string vnetName = "MYVM" + "_vnet";
string subnetName = "mySubnet";
string[] addressPrefixes = { "10.0.0.0/16" };
AddressSpace addressSpace = new AddressSpace(addressPrefixes);

VirtualNetworkData vnetData = new VirtualNetworkData()
{
    AddressSpace = addressSpace,
    Subnets =
    {
        new SubnetData()
        {
            Name = subnetName,
            AddressPrefix = "10.0.0.0/24"
        }
    }
};
VirtualNetwork vnet = (await resourceGroup.GetVirtualNetworks().CreateOrUpdateAsync(vnetName, vnetData)).Value;
```

> The main difference here is that a virtual network object is no longer needed to create a virtual network. One similarity is that subnets are defined inside virtual networks, however, with the new SDK you can get a subnets container using `.GetSubnets()`, and from there create any subnet in the virtual network from which the method is being called.

Reviewer: ```suggestion
In both libraries, subnets are defined inside virtual networks, however, with the new SDK you can get a subnets container using `.GetSubnets()`, and from there create any subnet in the virtual network from which the method is being called.
```

### Create a Security Group
#### Old
```C#
string nsgName = vmName + "_nsg";
NetworkSecurityGroup nsgParameters = new NetworkSecurityGroup()
{
    Location = location
};

NetworkSecurityGroup putNSgResponse = networkClient.NetworkSecurityGroups.CreateOrUpdate(rgName, nsgName, nsgParameters);
NetworkSecurityGroup nsg = networkClient.NetworkSecurityGroups.Get(rgName, nsgName);
```
#### New
```C#
string nsgName = vmName + "_nsg";
NetworkSecurityGroupData nsgData = new NetworkSecurityGroupData() { Location = location };
_ = resourceGroup.GetNetworkSecurityGroups().CreateOrUpdateAsync(nsgName, nsgData);
```

> Creating a Network security group does not longer require a `NetworkSecurityGroup` object as a parameter.

Reviewer: ```suggestion
Creating a Network security group no longer requires a `NetworkSecurityGroup` object as a parameter.
```

Reviewer: I think I would just remove this line - actually the structure of the models is very similar between the SDKs, the main difference is in operation signatures, and how the operations them selves are accessed.

    },
    IpConfigurations = new List<NetworkInterfaceIPConfiguration>()
    {
        new NetworkInterfaceIPConfiguration()
        {
            Name = ipConfigName,
            PrivateIPAllocationMethod = IPAllocationMethod.Dynamic,
            Subnet = subnetResponse,
        }
    },
    NetworkSecurityGroup = nsg
};

NetworkInterface putNicResponse = networkClient.NetworkInterfaces.CreateOrUpdate(rgName, nicname, nicParameters);
NetworkInterface nicResponse = networkClient.NetworkInterfaces.Get(rgName, nicname);
```
#### New
```C#
string nicname = vmName + "_nic";
NetworkInterfaceData nicData = new NetworkInterfaceData()
{
    Location = location,
    IpConfigurations = new List<NetworkInterfaceIPConfiguration>()
    {
        new NetworkInterfaceIPConfiguration()
        {
            Name = "Primary",
            Primary = true,
            Subnet = new Subnet() { Id = vnet.Subnets.First().Id },
            PrivateIPAllocationMethod = IPAllocationMethod.Dynamic,
            PublicIPAddress = new PublicIPAddress() { Id = ipAddress.Id }
        }
    }
};
NetworkInterface nic = (await networkInterfaceContainer.CreateOrUpdateAsync(nicName, nicData)).Value;
```

> This step is similar to the old SDK, however, notice that the `StartCreateOrUpdate()` method returns the network interface that has been created. 

Reviewer: ```suggestion
This step is similar to the old SDK, however, notice that the `CreateOrUpdateAsync()` method returns the network interface that has been created. 
```
